luarules/gadgets/unit_projectile_leash.lua

```diff
@@ -0,0 +1,163 @@
+-- use weaponDef customparams.projectile_leash_range to destroy projectiles when they exceed range/projectile_overrange_distance and projectile_leash_range relative to unit position.
+
+---- optional customParams ----
+-- use weaponDef customparams.weaponDef.customParams.projectile_destruction_method = "string" to change how projectiles are destroyed.
+-- "explode" (default if undefined) detonates the projectile.
+-- "descend" moves the projectile downward until it is destroyed by collision event.
+
+--static values
+local lazyUpdateFrames = math.ceil(Game.gameSpeed / 3 * 2)
+local edgyUpdateFrames = math.ceil(Game.gameSpeed / 6)
+local forcedDescentUpdateFrames = math.ceil(Game.gameSpeed / 5)
+local minimumThresholdRange = 10 --so that starburst missiles don't trigger edgyWatch during ascent
+local compoundingMultiplier = 1.1 --compounding multiplier that influences the arc at which projectiles are forced to descend
+local descentSpeedStartingMultiplier = 0.1
+
+--functions
+local spGetUnitPosition = Spring.GetUnitPosition
+local mathSqrt = math.sqrt
+local mathMax = math.max
+local spGetProjectilePosition = Spring.GetProjectilePosition
+local spSetProjectileCollision = Spring.SetProjectileCollision
+local spGetProjectileVelocity = Spring.GetProjectileVelocity
+local spSetProjectileVelocity = Spring.SetProjectileVelocity
+
+--tables
+local defWatchTable = {}
+local lazyProjectileWatch = {}
+local edgyProjectileWatch = {}
+local forcedDescentTable = {}
+
+
+for weaponDefID, weaponDef in pairs(WeaponDefs) do
+	if weaponDef.customParams.projectile_leash_range then
+		defWatchTable[weaponDefID] = {}
+		defWatchTable[weaponDefID].leashRange = tonumber(weaponDef.customParams.projectile_leash_range)
+	end
+	if weaponDef.customParams.projectile_leash_range or weaponDef.customParams.projectile_overrange_distance then
+		defWatchTable[weaponDefID] = defWatchTable[weaponDefID] or {}
+		defWatchTable[weaponDefID].range = weaponDef.range
+		defWatchTable[weaponDefID].weaponRange = weaponDef.range
+		defWatchTable[weaponDefID].overRange = tonumber(weaponDef.customParams.projectile_overrange_distance) or weaponDef.range
+		defWatchTable[weaponDefID].rangeThreshold = math.max((defWatchTable[weaponDefID].overRange - weaponDef.projectilespeed * lazyUpdateFrames), minimumThresholdRange)
+		defWatchTable[weaponDefID].weaponDefID = weaponDefID
+
+		local destructionMethod = weaponDef.customParams.projectile_destruction_method or "explode"
+		if destructionMethod == "descend" then
+			defWatchTable[weaponDefID].descentMethod = true
+		else
+			defWatchTable[weaponDefID].explodeMethod = true
+		end
+
+		Script.SetWatchWeapon(weaponDefID, true)
+	end
+end
+
+
+local function projectileOverRangeCheck(proOwnerID, weaponRange,  leashRange, originX, originZ, projectileX, projectileZ)
+	local dx1 = originX - projectileX
+	local dz1 = originZ - projectileZ
+	local distanceToOrigin = mathSqrt(dx1 * dx1 + dz1 * dz1)
```

Comment:
no need to sqrt here, just check against squared numbers, ie:

```LUA
local weaponRangeSq = weaponRange * weaponRange
local distanceToOriginSq = dx1 * dx1 + dz1 * dz1
if distanceToOriginSq > weaponRangeSq then ...
```

(same thing with distanceToOwner and leashRange)

Also since I think you only use weaponRange and leashRange you could just store weaponRangeSq and leashRangeSq since the beginning avoiding to recalculate them all the time

Reply:
donezo in the rewrite